File: binarysearch/binary_search.py
def binary_search(*, lst: list[int], x: int) -> list[int]:
    res = []
    start = 0;
    end = int(len(lst)) - 1
    mid = 0
    n = 0
    while start <= end:
        mid = int((end + start) / 2)
        n += 1
        print("============")
        print(f"lst[mid] is {lst[mid]}\nN is {n}")
        if x == lst[mid]:
            res.append(lst[mid])
            res.append(n)
            return res
        elif x > lst[mid]:
            start = mid + 1
        elif x < lst[mid]:
            end = mid - 1
    return None

def binary_search_recursive(*, lst: list[int], x: int, start: int, end: int, n: int) -> list[int]:
    mid = (end + start) // 2
    if lst[(end + start) // 2] != x and start <= end:
        if x > lst[mid]:
            return binary_search_recursive(lst=lst, x=x, start=(mid) + 1, end=end, n=n+1)
        elif x < lst[mid]:
            return binary_search_recursive(lst=lst, x=x, start=start, end=(mid) - 1, n=n+1)
    elif lst[mid] == x:
        if n == 0:
            n = 1
        res = res=[lst[mid], n]
        return res
    return None

lst = [x for x in range(100)]
finded_lst = binary_search_recursive(lst=lst, x=72, start=0, end=int(len(lst)), n=0)
finded_lst = binary_search(lst=lst, x=72)

print("============")
print (f"The x is {finded_lst[0]}")
print (f"Walk count is {finded_lst[1]}")

File: binarysearch/binary_search.c
#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>

int *binary_search(int *lst, int lst_size, int x)
{
    int start = 0;
    int end = lst_size - 1;
    int mid;
    int n = 0;
    int *res = malloc(sizeof(int) * 2);
    while(start <= end)
    {
        n++;
        mid = (start + end) / 2;
        printf("==================\n");
        printf("lst[mid] is %d\nN is %d\n", lst[mid], n);
        if (lst[mid] == x)
        {
            res[0] = lst[mid];
            res[1] = n;
            return res;
        }
        else if (x > lst[mid])
            start = mid + 1;
        else if (x < lst[mid])
            end = mid - 1;
    }
    return NULL;
}

int main(int ac, char **av)
{
    int *lst = malloc(100 * sizeof(int));
    for (int i = 0; i < 100; i++)
        lst[i] = i;
    int *finded_lst = binary_search(lst, 100, atoi(av[1]));
    printf("==================\n");
    printf("The x is %d\nWalk Count is %d", finded_lst[0], finded_lst[1]);
    free(lst);
    free(finded_lst);
    return 0;
}
